File: manila/sources.py
"""Where a folder's news arrives.

Manila never reads a mailbox or a Teams channel. What it can do is remember
which ones a folder is about, so an assistant that *can* read them is told
where to look instead of guessing from the folder's name -- and so that two
two people's idea of what a folder covers is one list you can see rather than
a sentence in a prompt somewhere.

What goes in here is pointers and nothing else: a team and channel id, a chat
id, an address, a domain. Not keywords. The distinction is what keeps the list
honest. Ids change perhaps twice a year and you notice when they do, because
the sync comes back empty. Keywords change every week, and a stale keyword
list fails quietly -- it keeps working, just on last quarter's vocabulary. The
words worth searching for are already written down in the folder's own Items
and Waiting On cells, where they are kept current by being used, so that is
where they should be read from.

Nothing here is written by an assistant. The agent route table can read this
list and cannot add to it, the same as everything else in a folder.
"""

import re
import urllib.parse

from . import db, history

# What a pointer can be. Each kind names the one field that identifies it;
# `within` is filled only where the id is not enough on its own, since a Graph
# channel id means nothing without the team that holds it. `add` and `help` are
# what the box asks for when you are adding one, which is a different question
# from what the value is called once it is stored.
KINDS = {
    "teams_channel": {
        "label": "Teams channel",
        "ref": "channel id",
        "within": "team id",
        "add": "paste the channel link",
        "help": "In Teams, right-click the channel and choose Get link to "
                "channel. The channel id, the team that holds it, and the "
                "channel's own name all come out of that one link.",
    },
    "teams_chat": {
        "label": "Teams chat",
        "ref": "chat id",
        "add": "chat id, or the chat's web address",
        "help": "Chats have no copy-link. Open the chat at teams.microsoft.com "
                "and copy the address bar -- the 19:... in it is the chat id.",
    },
    "email_folder": {
        "label": "Mail folder",
        "ref": "folder path",
        "add": "Inbox/Clients/Acme",
        "help": "Mail you have filed into this Outlook folder, or into any "
                "folder beneath it. Write the path as the folder pane shows "
                "it, from the top, with / between the levels.",
    },
    "email_address": {
        "label": "Email address",
        "ref": "address",
        "add": "someone@example.com",
        "help": "Mail from this person, wherever in their organization they sit.",
    },
    "email_domain": {
        "label": "Email domain",
        "ref": "domain",
        "add": "example.com",
        "help": "Mail from anyone at this domain.",
    },
}

# Addresses and domains are matched by whoever reads the mailbox, and no mail
# system treats case as meaningful in either. Ids are opaque and are left
# exactly as they were pasted.
FOLDED = ("email_address", "email_domain")

# A mail folder keeps the case it was typed in, because that is how it is
# shown back to you, but Outlook will not hold "Acme" and "acme" side by
# side, so two paths differing only in case are the same folder.
CASELESS = ("email_folder",)


# --- reading a pointer out of the link you were given -------------------------
#
# Nobody has a channel id. What they have is the link Teams hands out, which is
# where that id is written down in a form you can copy -- percent-encoded, with
# the team id beside it as `groupId` and the channel's name in between. Asking
# for the link is asking for the thing that actually exists; asking for the two
# ids is asking someone to do a URL decode by hand, and getting `19%3a...` in
# the box instead, which matches nothing and says nothing about why.

CHANNEL_IN_LINK = re.compile(r"/l/channel/([^/?#]+)(?:/([^/?#]*))?")
CHAT_IN_LINK = re.compile(r"19:[^/?#&\s]+")


def _is_link(text):
    return text.startswith(("http://", "https://")) or "/l/channel/" in text


def channel_from_link(url):
    """The channel id, the team that holds it, and its name, out of one link."""
    parts = urllib.parse.urlsplit(url)
    found = CHANNEL_IN_LINK.search(parts.path) or CHANNEL_IN_LINK.search(parts.fragment)
    if not found:
        raise ValueError(
            "that does not look like a channel link. In Teams, right-click the "
            "channel and choose Get link to channel, then paste the whole thing.")
    query = urllib.parse.parse_qs(parts.query) | urllib.parse.parse_qs(
        urllib.parse.urlsplit(parts.fragment).query)
    group = next((v[0].strip() for k, v in query.items() if k.lower() == "groupid"), "")
    if not group:
        # A link copied from the address bar rather than from the menu can come
        # without it, and a channel id on its own is not enough to find a channel.
        raise ValueError(
            "that link carries no groupId, so it does not say which team the "
            "channel is in. Use Get link to channel on the channel itself.")
    return {"ref": urllib.parse.unquote(found.group(1)),
            "within": group,
            "name": urllib.parse.unquote(found.group(2) or "").strip()}


def chat_from_link(url):
    """The chat id out of a chat's web address, which is the only place it shows."""
    found = CHAT_IN_LINK.search(urllib.parse.unquote(url))
    if not found:
        raise ValueError(
            "no chat id in that. Open the chat at teams.microsoft.com and copy "
            "the address bar -- the part beginning 19: is the chat id.")
    return {"ref": found.group(0)}


def _unpack(kind, ref):
    """What a pasted link says, for the kinds that come as links."""
    if not _is_link(ref):
        return {}
    if kind == "teams_channel":
        return channel_from_link(ref)
    if kind == "teams_chat":
        return chat_from_link(ref)
    return {}


def folder_path(text):
    """A mail folder path, with its separators and stray spaces made regular.

    Backslashes are accepted because that is how a Windows user writes a path,
    and the result always uses /, so the same folder typed two ways is stored
    one way.
    """
    parts = [part.strip() for part in re.split(r"[/\\]", text)]
    return "/".join(part for part in parts if part)


def _clean(body):
    """One incoming source, checked into the row that would store it."""
    kind = str(body.get("kind") or "").strip()
    if kind not in KINDS:
        raise ValueError(f"kind must be one of {', '.join(KINDS)}, not {kind!r}")
    spec = KINDS[kind]

    ref = str(body.get("ref") or "").strip()
    if not ref:
        raise ValueError(f"a {spec['label'].lower()} needs its {spec['ref']}")

    # A pasted link is unpacked into the ids it carries, and anything it also
    # happens to know -- a channel's name -- is used only where you have not
    # said otherwise.
    from_link = _unpack(kind, ref)
    ref = from_link.get("ref", ref)
    if kind in FOLDED:
        ref = ref.lower().lstrip("@")
    if kind == "email_folder":
        ref = folder_path(ref)
        if not ref:
            raise ValueError("a mail folder needs its path, e.g. "
                             "Inbox/Clients/Acme")

    within = str(body.get("within") or "").strip() or from_link.get("within", "")
    if spec.get("within") and not within:
        raise ValueError(
            f"a {spec['label'].lower()} needs its {spec['within']} too, and a "
            "channel id does not carry one. Paste the channel link instead: "
            "right-click the channel in Teams -> Get link to channel.")
    if not spec.get("within"):
        within = ""

    return {"kind": kind, "ref": ref, "within": within,
            "name": str(body.get("name") or "").strip() or from_link.get("name", "")}


def shape(row):
    out = {"id": row["id"], "kind": row["kind"], "name": row["name"],
           "ref": row["ref"], "within": row["within"]}
    if row["kind"] == "email_folder":
        # Mail search tools take the folder's own name, not its path. The path
        # stays the pointer: two folders can share a leaf, and it is the path
        # that tells a result from one apart from the other.
        out["leaf"] = row["ref"].rsplit("/", 1)[-1]
    return out


def listing(conn, folder_id):
    rows = conn.execute(
        "SELECT * FROM folder_source WHERE folder_id = ? ORDER BY position, id",
        (folder_id,)).fetchall()
    return {"sources": [shape(row) for row in rows]}


def _key(kind, ref):
    return ref.casefold() if kind in CASELESS else ref


def _clash(conn, folder_id, entry, excluding=None):
    """The same pointer twice in one folder is a mistake, not a preference."""
    found = conn.execute(
        "SELECT id, kind, ref FROM folder_source WHERE folder_id = ? AND kind = ?",
        (folder_id, entry["kind"])).fetchall()
    want = _key(entry["kind"], entry["ref"])
    return any(row["id"] != excluding and _key(row["kind"], row["ref"]) == want
               for row in found)


def add(conn, folder_id, body):
    entry = _clean(body)
    if _clash(conn, folder_id, entry):
        raise ValueError("this folder already listens to that one")
    row_id = conn.execute(
        "INSERT INTO folder_source (folder_id, kind, name, ref, within, position,"
        " added_at) VALUES (?, ?, ?, ?, ?, ?, ?)",
        (folder_id, entry["kind"], entry["name"], entry["ref"], entry["within"],
         db.next_position(conn, "folder_source", folder_id), history.now()),
    ).lastrowid
    return shape(conn.execute("SELECT * FROM folder_source WHERE id = ?",
                              (row_id,)).fetchone())


def _source(conn, source_id):
    row = conn.execute("SELECT * FROM folder_source WHERE id = ?",
                       (source_id,)).fetchone()
    if row is None:
        raise KeyError(f"no source with id {source_id}")
    return row


def update(conn, source_id, body):
    """Change one. The kind is fixed -- a channel does not become a domain."""
    row = _source(conn, source_id)
    entry = _clean({"kind": row["kind"],
                    "ref": body.get("ref", row["ref"]),
                    "within": body.get("within", row["within"]),
                    "name": body.get("name", row["name"])})
    if _clash(conn, row["folder_id"], entry, excluding=source_id):
        raise ValueError("this folder already listens to that one")
    conn.execute(
        "UPDATE folder_source SET name = ?, ref = ?, within = ? WHERE id = ?",
        (entry["name"], entry["ref"], entry["within"], source_id))
    return shape(_source(conn, source_id))


def remove(conn, source_id):
    _source(conn, source_id)
    conn.execute("DELETE FROM folder_source WHERE id = ?", (source_id,))
    return {"id": source_id, "deleted": True}

File: manila/static/sources.js
/* Sources: where this folder's news arrives.
 *
 * Pointers and nothing else -- a team and channel id, a chat id, an address, a
 * domain. Not keywords. An id changes about twice a year and tells you when it
 * has, because the sync comes back empty; a keyword list goes stale quietly and
 * keeps working on last quarter's vocabulary. The words worth searching for are
 * already in the folder's own Items and Waiting On cells, kept current by being
 * used, so that is where an assistant reads them from.
 *
 * It sits above the queue because it is the other half of the same question:
 * this says where to look, and the queue below says what came back.
 */

import { api, toast } from "./api.js";
import * as fmt from "./format.js";
import * as icons from "./icons.js";
import { el } from "./table.js";

export class Sources {
  constructor(host, { folderId, kinds }) {
    Object.assign(this, { host, folderId, kinds });
    this.sources = [];
    this.runs = [];
    this.adding = null;      // the kind being added, while the row is open
  }

  /* Remembered if you have ever said, and otherwise open exactly when the list
   * is empty. A folder with nothing listed is one where the panel has
   * something to teach; a folder already set up wants its queue at the top of
   * the screen, not a list of ids it will not touch again this quarter. */
  get open() {
    const choice = localStorage.getItem("manila.sources");
    if (choice === "open") return true;
    if (choice === "shut") return false;
    return this.sources.length === 0;
  }

  path(rest = "") {
    return api.p(`/folders/${this.folderId}/sources${rest}`);
  }

  async load() {
    const payload = await api.get(this.path());
    this.sources = payload.sources ?? [];
    this.runs = payload.last_runs ?? [];
  }

  async refresh() {
    await this.load();
    this.render();
  }

  spec(kind) {
    return this.kinds[kind] ?? { label: kind, ref: "id" };
  }

  // --- rendering ---------------------------------------------------------------

  render() {
    this.host.innerHTML = "";
    const section = el("section", "section sources");

    const head = el("div", "section-head");
    const shown = this.open;
    const toggle = el("button", "sources-toggle");
    toggle.append(el("span", "twist", shown ? "▾" : "▸"),
                  el("span", "section-title", "Sources"));
    toggle.onclick = () => {
      localStorage.setItem("manila.sources", shown ? "shut" : "open");
      this.render();
    };
    head.append(toggle);
    if (!shown) head.append(el("span", "sources-line", this.summary()));
    section.append(head);

    if (shown) section.append(this.panel());
    this.host.append(section);
  }

  /* Shut, it still has to be worth glancing at: what this folder listens to,
   * and whether anything has actually been read from it. */
  summary() {
    if (!this.sources.length) return "none set";
    const counts = new Map();
    for (const source of this.sources) {
      counts.set(source.kind, (counts.get(source.kind) ?? 0) + 1);
    }
    const parts = [...counts].map(([kind, n]) =>
      `${n} ${this.spec(kind).label}${n === 1 ? "" : "s"}`);
    return `${parts.join(", ")} · ${this.read()}`;
  }

  read() {
    if (!this.runs.length) return "never read";
    return this.runs.map((run) => {
      const point = run.covered_through || run.last_run;
      const word = run.covered_through ? "read to" : "last run";
      return `${run.source} ${word} ${fmt.stamp(point)}`;
    }).join(" · ");
  }

  panel() {
    const box = el("div", "sources-box");
    box.append(el("p", "hint",
      "Where an assistant should start looking for this folder's news: Teams "
      + "channels and chats by id, mail by the folder you file it in, or by "
      + "address or domain. Pointers only -- what "
      + "to search for is read off this folder's own items, which stay current "
      + "because you use them. A starting point, not a fence: internal mail and "
      + "places nobody has listed are still read, which is how a task this "
      + "folder does not mention yet reaches the queue at all."));

    if (this.sources.length) {
      const list = el("div", "source-list");
      for (const source of this.sources) list.append(this.row(source));
      box.append(list);
    }

    box.append(this.adding ? this.addCard() : this.addBar());
    box.append(this.coverage());
    return box;
  }

  row(source) {
    const spec = this.spec(source.kind);
    const row = el("div", "source-row");
    row.append(el("span", "source-kind", spec.label));
    row.append(this.field(source, "name", "what you call it", "source-name"));
    const ref = this.field(source, "ref", spec.ref, "source-ref");
    row.append(ref);
    if (spec.within) {
      row.append(this.field(source, "within", spec.within, "source-within"));
    } else {
      ref.classList.add("whole");
    }

    const drop = el("button", "icon trash");
    drop.innerHTML = icons.trash;
    drop.title = "Stop listening to this one";
    drop.setAttribute("aria-label", "Remove this source");
    drop.onclick = async () => {
      const what = source.name || source.ref;
      if (!confirm(`Stop reading ${what} into this folder?`)) return;
      try {
        await api.post(api.p(`/sources/${source.id}/delete`));
        await this.refresh();
      } catch (error) { toast(error.message); }
    };
    row.append(drop);
    return row;
  }

  /* Saved when you leave the box, not as you type: these are ids pasted in one
   * go, and a PATCH per keystroke would send a dozen half-ids. */
  field(source, key, placeholder, className) {
    const input = el("input", className);
    input.value = source[key] ?? "";
    input.placeholder = placeholder;
    input.spellcheck = false;
    input.onchange = async () => {
      const value = input.value.trim();
      if (value === (source[key] ?? "")) return;
      try {
        const { source: saved } = await api.patch(
          api.p(`/sources/${source.id}`), { [key]: value });
        Object.assign(source, saved);
        input.value = saved[key];
      } catch (error) {
        toast(error.message);
        input.value = source[key] ?? "";     // put back what is actually stored
      }
    };
    return input;
  }

  addBar() {
    const bar = el("div", "source-add");
    for (const [kind, spec] of Object.entries(this.kinds)) {
      const button = el("button", "ghost", `Add ${spec.label}`);
      button.onclick = () => {
        this.adding = { kind, ref: "" };
        this.render();
      };
      bar.append(button);
    }
    return bar;
  }

  /* One box, because there is one thing to paste. A channel link carries the
   * channel id, the team that holds it and the channel's name, and asking for
   * those three separately is asking someone to do a URL decode by hand.
   *
   * It is a card rather than a row in the list above: the list is five columns
   * wide and this is a question with an explanation and two answers, which is
   * not the same shape. The previous version tried to be a row, overflowed the
   * grid, and put Cancel somewhere you could not see it. */
  addCard() {
    const draft = this.adding;
    const spec = this.spec(draft.kind);
    const card = el("div", "source-new");
    card.append(el("div", "source-new-kind", spec.label));

    const input = el("input", "source-new-ref");
    input.placeholder = spec.add ?? spec.ref;
    input.spellcheck = false;
    input.oninput = () => { draft.ref = input.value; };
    input.onkeydown = (event) => {
      if (event.key === "Enter") save();
      if (event.key === "Escape") this.stopAdding();
    };
    card.append(input);
    if (spec.help) card.append(el("p", "hint", spec.help));

    const save = async () => {
      if (!(draft.ref ?? "").trim()) return input.focus();
      try {
        await api.post(this.path(), { kind: draft.kind, ref: draft.ref });
        this.adding = null;
        await this.refresh();
      } catch (error) { toast(error.message); }
    };

    const actions = el("div", "source-new-actions");
    const add = el("button", "primary", "Add");
    add.onclick = save;
    const cancel = el("button", "ghost", "Cancel");
    cancel.onclick = () => this.stopAdding();
    actions.append(add, cancel);
    card.append(actions);

    Promise.resolve().then(() => input.focus?.());
    return card;
  }

  stopAdding() {
    this.adding = null;
    this.render();
  }

  /* How far each source has actually been read, which is the fact that decides
   * how far back the next run goes. A source with no run at all is the one
   * worth seeing: a first pass over it has no floor, so it will surface a
   * backlog rather than a morning unless it is given a window. */
  coverage() {
    const box = el("div", "source-coverage");
    if (!this.runs.length) {
      box.append(el("span", "muted", "Nothing has been read into this folder yet."));
      return box;
    }
    for (const run of this.runs) {
      const line = el("div", "coverage-run");
      line.append(el("span", "coverage-source", run.source));
      if (run.covered_through) {
        line.append(el("span", "coverage-point",
          `read through ${fmt.stamp(run.covered_through)}`));
      } else {
        line.append(el("span", "coverage-point muted", "no watermark"));
      }
      const when = el("span", "coverage-when",
        `${run.runs} run${run.runs === 1 ? "" : "s"}, last ${fmt.stamp(run.last_run)}`);
      line.append(when);
      box.append(line);
    }
    return box;
  }
}
